man/man1/get.1
.\" Written by Patrick Hochstenbach <hochsten@lanl.gov>
.TH get 1 "5 December 2002" "get 0.0.1" "get manual"
.SH NAME
get \- a trivial WWW user agent
.SH SYNOPSIS
.B get [options] http_address
.SH DESCRIPTION
This command is part of the srepod package and is used to fetch static
repositories from the network. It is a stripped down implementation
of the curl interface.
.SH OPTIONS
.IP "-h"
Include the header in the ourput
.IP "-p"
Try to bypass proxy servers
.IP "-o file"
Write the output to this file.
.IP "-s maxsize"
Maximum number of bytes to read from the network.
.IP "-t seconds"
Maximum number of seconds this program can run.
.SH "EXIT VALUES"
.IP 0
Success
.IP 1
Invalid command line arguments.
.IP 2
Output file can't be created.
.IP 3
Timeout value reached.
.IP 4
Maxsize value reached.
.IP 5
Internal error.
.SH "SEE ALSO"
.BR curl (1),
.BR GET (1),
.BR wget (1)
.SH BUGS
No bugs known at this date. Please mail me <hochsten@lanl.gov> if you find some.


ChangeLog
Tue Dec 19 09:17:53 CET 2006  Patrick.Hochstenbach@ugent.be
	* Minor bug fixes
    * Maintenance release
	* Tagged the code sr2_1_6

Tue Jul 26 09:56:47 CEST 2005 Patrick.Hochstenbach@ugent.be
        * Changed layout HTML pages
        * Change environ file to have correct version numbers
          updated gateway locations to neutral 'localhost'
	    * Fixed Makefile to do a better 'make clean'
        * Changed the UPGRADE procedure
    	* Tagged the code sr2_1_5

Mon Jun 27 14:36:05 CEST 2005 Patrick.Hochstenbach@ugent.be
    	* Fixed namespace program; namespace declarations and 
      	  schemaLocations shouldn't be overwritten in metadata 
      	  childs if they already exist 
      	  [bug report from haejoong@ldc.upen.edu]
    	* Tagged the code sr2_1_4

Tue Jun  7 20:10:17 CEST 2005 Patrick.Hochstenbach@ugent.be
    	* Fixed Last-Modified bug, should be case-insensitive
      	  [bug report from haejoong@ldc.upen.edu]
    	* Tagged the code sr2_1_3

Thu Apr  7 11:04:43 CEST 2005 Patrick.Hochstenbach@ugent.be
    	* Fixed buffer initalization bug in scan2.c [bug
      	  report from haejoong@ldc.upen.edu]
    	* Fixed parse failed bug in transform.c [bug report
      	  from haejoong@ldc.upen.edu]
    	* Tagged the code sr2_1_2

Sat Jan 22 11:29:48 CET 2005 Patrick.Hochstenbach@ugent.be
    	* Fixed the validurl program 
    	* Tagged the code sr2_1_1

Fri Apr  9 14:03:23 MDT 2004 hochsten@sci.kun.nl
	* Added AUTHORS to packages
	* Tagged the code sr2_1_0

Wed Apr  7 13:33:26 MDT 2004 hochsten@sci.kun.nl
	* Created INSTALL.html file

Wed Mar 31 15:19:13 MST 2004 hochsten@sci.kun.nl
	* Added terminator.cgi script
	* Added termination support in the create_cache.sh
	* Added optional termination and termination webpages
	  to environ
	* Updated documentation

Sat Mar 20 13:19:06 MST 2004 hochsten@lanl.gov
	* Added INSTALL to the $(INSTDIR)
	* The gateway optionally uses an error.html template if the
	  environmental variable GWERRORHTML is set
	* Added support for the 'terminate' verb.
	
Sun Mar 14 15:48:17 MST 2004 hochsten@lanl.gov
	* Added validurl binary
	* Added (optionally) checking of valid baseURL's
	  in static repository files [the check is mandatory in
	  the beta release of the protocol specs]

Sat Mar 13 08:59:57 MST 2004 hochsten@lanl.gov
	* Deleted incorrect /tmp install 
	* Edited INSTALL file
	* Added chmod in registry.cgi, this way the friends.conf
	  keeps the correct file permissions during registration
	* Added functionality to the cconfig package to write
	  changed configuration files back to disk
	* Added support to delete (unregister) a static repository
	  from the gateway

Mon Dec 22 12:49:43 MST 2003 hochsten@lanl.gov
	* Fixed the order in desc.conf

Thu Dec  4 15:01:23 MST 2003 hochsten@lanl.gov
	* Debugged the installation scripts so that installations
	  in a non-default directory is somewhat easier.
	* Tagged the code sr2_0_4  

Tue Dec  2 16:59:15 MST 2003 hochsten@lanl.gov
	* Added more elaborate error messages so that Internet Explorer
	  users can see them.
	* Tagged the code sr2_0_3

Fri Oct 31 14:14:28 MST 2003 hochsten@lanl.gov
	* Changed the timeout value for the up-to-date test to
	  60 seconds.

Thu Oct 30 21:25:23 MST 2003 hochsten@lanl.gov
	* Fixed buffer overflow check in scan2.c.
 
Thu Oct 30 20:29:37 MST 2003 hochsten@lanl.gov
	* Fixed bug in schemaLocation, to many whitespace characters
	  were eaten up during the Static Repository transformation.
	* Tagged the code sr2_0_2.

Tue Oct 14 16:35:36 MDT 2003 hochsten@lanl.gov
	* The schema of the gateway description has changed.
          A URL element inside gatewayDescription is not anymore
	  needed.
	* Tagged the code sr2_0_1.

Tue Oct 14 13:56:05 MDT 2003 hochsten@lanl.gov
	* Added support for the beta release of the static 
   	  repository gateway specitication.
	* Deleted support of alpha version registration process.
	  The administrator of a static repository has to send
	  an initiate request to the base URL of the gateway to
	  start the registration process.
	* Made the registration procedure more configurable. In the
	  environ file you can specify which type of confirmation
	  messages you want to send to the static repository 
	  administrator before starting the registration process.
	* Tagged the code sr2_0_0.
	   
Tue Sep 16 13:05:42 MDT 2003 hochsten@lanl.gov
	* Fixed a bug in the hstrip function which prevented
	  Static Repositories created with a '\r' eol parsed
	  correctly.
	* Tagged the code sr1_0_26  
	
Tue Sep 16 12:17:01 MDT 2003 hochsten@lanl.gov
	* Fixed bug in the fixed code of Jul 3 which prevented
	  metadata records been written to the dbf file
	* Tagged the code sr1_0_25  

Thu Jul  3 08:48:09 MDT 2003 hjerez@lanl.gov
	* Fixed a bug in the transform_c code. Got corrupted
	  databases if oai namespace elements were used in 
	  the namespace of records

Mon Jun 30 10:42:11 MDT 2003 hochsten@lanl.gov

	* Fixed bug in namespace code, got a segmentation fault
	  on empty namespace declarations.
	* Added three new example repositories: vbns.xml vbns2.xml vbns3.xml
	* Tagged the code sr1_0_23

Tue Jun 24 12:07:40 MDT 2003 hochsten@lanl.gov

	* Added LICENSE file, put code under BSD license
	* Tagged the code sr1_0_22alpha
	
Tue Jun 24 09:19:37 MDT 2003 hochsten@lanl.gov

	* The LAUR # 03-4183 has been assigned to:
	  A C Implementation of an OAI-PMH Static Repository Gateway
	* Tagged the code sr1_0_21alpha 

Thu Apr  3 10:54:01 MST 2003 hochsten@lanl.gov

	* Put some extra buffer overflow checks in
	  the namespace code

Wed Apr  2 14:26:06 MST 2003 hochsten@lanl.gov

	* Fixed a bug in namespace code
	* Created relative paths to the environ
	  script in create_cache.sh. The srepod
	  child will chdir to the spool directory
	  at startup
	* Tagged the code sr1_0_19alpha  

Wed Apr  2 08:49:20 MST 2003 hochsten@lanl.gov

	* Changed desc.conf, the admin email is now 
	  configurable in the environ file
	* Changed srepod.conf, the pathnames are not

src/daemon/get.c
/* 
 * Authors: Patrick Hochstenbach <Patrick.Hochstenbach@UGent.be>
 *
 * $Id: get.c,v 1.2 2006/12/19 08:33:14 hochstenbach Exp $
 * 
 * get - fetch data from a http address on the Internet
 *
 * usage: get [options] http_address
 *
 * options:
 *	-h 	      (include the head in the output)
 *	-p	      (try to bypass proxy servers)
 * 	-o file
 * 	-s maxsize
 * 	-t timeout
 * 	
 * exit values:
 *
 * 	0 	- success
 * 	1	- invalid command arguments
 * 	2	- output file can't be created
 * 	3	- timeout reached
 * 	4	- maxsize reached
 * 	5	- internal error
 *
 * only exit value 0 will create an output file (except when writing to stdout).	
 *
 */
#include <stdio.h>
#include <curl/curl.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define	MAX_PATH 1024
#define MAX_URL	 1024
#define BUF_SIZE 1024

extern char 	*optarg;
extern int 	optind, opterr, optopt;

typedef struct {
	FILE	*fp;
	int	timeout;
	int	maxsize;
	size_t	cursize;
	char	outfile[MAX_PATH];
} opt_struct;

void	usage(void);
size_t 	mywriter(void *ptr, size_t size, size_t nmemb, void *userp);

void	usage(void) {
	fprintf(stderr,"usage: get [options] http_address\n"
		       "\n"
		       "options:\n"
		       "        -h\n"
		       "        -p\n"
		       "        -o file\n"
		       "        -s maxsize\n"
		       "        -t timeout\n\n"
	       );
	exit(1);
}

size_t 	mywriter(void *ptr, size_t size, size_t nmemb, void *userp) {
	size_t 	   n;
	opt_struct *opt_p;

	opt_p = userp;

	/* Check if we don't get more than we requested */
	if (opt_p->maxsize > 0  && opt_p->cursize + size * nmemb > opt_p->maxsize) 
		return(-1);
	
	n = fwrite(ptr, size, nmemb, opt_p->fp);

	opt_p->cursize += n * size;

	return(n * size);
}
		       

int	main(int argc, char *argv[]) {
	int		exit_val;
	int		c;
	int		doheader = 0;
	int		doproxy  = 0;
	time_t		t;
	char		url[MAX_URL];
	CURL		*curl;
	CURLcode	res;
	opt_struct	opt;	

	opt.fp	       =  stdout; /* Default write to stdout */
	opt.timeout    = -1;
	opt.maxsize    = -1;
	opt.cursize    =  0;
	opt.outfile[0] = '\0';

	while( (c = getopt(argc,argv,"ho:ps:t:")) != EOF ) {
		switch (c) {
			case 'h':
				doheader = 1;
				break;
			case 'o':
				strncpy(opt.outfile,optarg,1024);
				/* make sure the filename has a null at the end */
				opt.outfile[1023] = '\0';
				break;
			case 'p':
				doproxy = 1;
				break;
			case 's':
				opt.maxsize = atoi(optarg);
				break;
			case 't':
				opt.timeout = atoi(optarg);
				break;
			default:
				usage();
		}
	}

	if (argc - optind == 1) {
		strncpy(url,argv[optind],1024);
		url[1023] = '\0';
	}
	else {
		usage();
	}

	if (doproxy) {
		time(&t);
		srand(t);
		sprintf(url, "%s?%d", url, rand());
	}

	curl = curl_easy_init();

	if (curl) {
		curl_easy_setopt(curl, CURLOPT_URL, url);

		curl_easy_setopt(curl, CURLOPT_HEADER, doheader);

		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, mywriter);
		curl_easy_setopt(curl, CURLOPT_FILE, &opt);

		if (opt.timeout > 0) 
			curl_easy_setopt(curl, CURLOPT_TIMEOUT, opt.timeout);

		if (opt.outfile[0] != '\0') {
			umask(S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);

			opt.fp = fopen(opt.outfile, "w");

			if (opt.fp == NULL) {
				exit(2);
			}
		}

		res = curl_easy_perform(curl);

		fclose(opt.fp);

		if (res == CURLE_OK) {
			exit_val = 0;
		}	
		else if (res == CURLE_OPERATION_TIMEOUTED) {
			unlink(opt.outfile);
			exit_val = 3;
		}
		else if (res == CURLE_WRITE_ERROR) {
			unlink(opt.outfile);
			exit_val = 4;
		}
		else {
			if (opt.outfile[0] != '\0')
				unlink(opt.outfile);
			exit_val = 5;
		}

		curl_easy_cleanup(curl);
	}
	else {
		exit_val = 5;
	}

	return(exit_val);
}
